feat(SBOM): Support SBOM generation

simar7 opened the pull request:
This adds support for Trivy's SBOM generation.

Signed-off-by: Simar <simar@linux.com>

itaysk commented on README.md:
```diff
@@ -227,6 +227,42 @@ jobs:
+### Using Trivy to generate SBOM
+It's possible for Trivy to generate an SBOM of your dependencies and submit them to a consumer like GitHub Dependency Snapshot.
+
+The sending of SBOM to GitHub feature is only available if you currently have [GitHub Dependency Snapshot](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-the-dependency-graph) available to you in your repo. 
```
probably this is a more accurate link? https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/using-the-dependency-submission-api

itaysk commented on README.md:
```diff
@@ -227,6 +227,42 @@ jobs:
+
+The sending of SBOM to GitHub feature is only available if you currently have [GitHub Dependency Snapshot](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-the-dependency-graph) available to you in your repo. 
+
+In addition to send results to the GitHub Dependency Snapshot, you will need to create a [GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
```
```suggestion
In order to send results to the GitHub Dependency Snapshot, you will need to create a [GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
```

simar7 replied:
@knqyf263 - `GITHUB_TOKEN` doesn't have the `repo` scope as @achton mentioned. Here are the docs https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

Here's a failed run to demo it: https://github.com/simar7/package-test/runs/6972488098?check_suite_focus=true

simar7 replied:
@knqyf263 - thanks for the idea, I like it. I've implemented like you mentioned, here's an example:

```
      - name: Run Trivy in GitHub SBOM mode and upload results
        uses: aquasecurity/trivy-action@master
        with:
          scan-type: 'sbom'
          format: 'github'
          output: 'dependency-results.sbom.json'
          artifact-type: 'fs'
          image-ref: '.'
          github-pat: ${{ secrets.PAT_TOKEN }}
```

krol3 reviewed (approved):
LGTM

knqyf263 commented on README.md:
```diff
@@ -227,6 +227,42 @@ jobs:
+
+      - name: Upload Trivy SBOM results to GitHub Dependency tab
+        run: |
+            curl -u "${{ secrets.PAT_TOKEN }}" -H 'Content-Type: application/json' 'https://api.github.com/repos/'$GITHUB_REPOSITORY'/dependency-graph/snapshots' -d @./dependency-results.sbom.json
```
What if we add this line in entrypoint.sh?

```
if [[ "${format}" == "github" ]]; then
  echo "Uploading GitHub Dependency Snapshot"
  curl -u "${{ secrets.PAT_TOKEN }}" -H 'Content-Type: application/json' 'https://api.github.com/repos/'$GITHUB_REPOSITORY'/dependency-graph/snapshots' -d @./dependency-results.sbom.json
fi
```

knqyf263 replied:
BTW, do we need PAT_TOKEN? Doesn't `secrets.GITHUB_TOKEN` work?

achton replied:
@knqyf263 As far as I can tell from [this page](https://docs.github.com/en/rest/dependency-graph/dependency-submission#create-a-snapshot-of-dependencies-for-a-repository) ("You must authenticate using an access token with the `repo` scope to use this endpoint for a repository that the requesting user has access to.") the `GITHUB_TOKEN` will not cut it. 
But I may be mistaken 🤷🏻

simar7 replied:
Default behaviour is currently documented with default value being 0 which is also the Trivy default.

knqyf263 replied:
Got it. Thanks for clarifying. 
It looks a bit weird. I believe GitHub will improve it so that GITHUB_TOKEN can be used, but ok, let's keep it at the moment.

achton commented on action.yaml:
```diff
@@ -68,7 +67,6 @@ inputs:
   hide-progress:
     description: 'hide progress output'
     required: false
-    default: 'true'
```
I'm not sure I understand the upstream default here:

```sh
$ trivy i --help | grep -i progress
   --no-progress                    suppress progress bar (default: false) [$TRIVY_NO_PROGRESS]
```

Is the default to display the progress bar if `--no-progress` is omitted? In that case, we would be breaking BC in the action, right?

If not, then fine ;) Just remember to update README as well.

simar7 replied:
Good catch - I had to make this change because the default behaviour of the action is to add `--no-progress` like you mentioned but the `sbom` mode in Trivy doesn't have this flag just yet. 

```
trivy sbom --help
NAME:
   trivy sbom - generate SBOM for an artifact

USAGE:
   trivy sbom [command options] ARTIFACT

DESCRIPTION:
   ARTIFACT can be a container image, file path/directory, git repository or container image archive. See examples.

OPTIONS:
   --artifact-type value, --type value  input artifact type (image, fs, repo, archive) (default: "image") [$TRIVY_ARTIFACT_TYPE]
   --clear-cache, -c                    clear image caches without scanning (default: false) [$TRIVY_CLEAR_CACHE]
   --db-repository value                OCI repository to retrieve trivy-db from (default: "ghcr.io/aquasecurity/trivy-db") [$TRIVY_DB_REPOSITORY]
   --ignorefile value                   specify .trivyignore file (default: ".trivyignore") [$TRIVY_IGNOREFILE]
   --insecure                           allow insecure server connections when using SSL (default: false) [$TRIVY_INSECURE]
   --offline-scan                       do not issue API requests to identify dependencies (default: false) [$TRIVY_OFFLINE_SCAN]
   --output value, -o value             output file name [$TRIVY_OUTPUT]
   --sbom-format value, --format value  SBOM format (cyclonedx, spdx, spdx-json, github) (default: "cyclonedx") [$TRIVY_SBOM_FORMAT]
   --severity value, -s value           severities of vulnerabilities to be displayed (comma separated) (default: "UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL") [$TRIVY_SEVERITY]
   --skip-dirs value                    specify the directories where the traversal is skipped  (accepts multiple inputs) [$TRIVY_SKIP_DIRS]
   --skip-files value                   specify the file paths to skip traversal                (accepts multiple inputs) [$TRIVY_SKIP_FILES]
   --timeout value                      timeout (default: 5m0s) [$TRIVY_TIMEOUT]
   
EXAMPLES:
  - image scanning:
      $ trivy sbom alpine:3.15

  - filesystem scanning:
      $ trivy sbom --artifact-type fs /path/to/myapp

  - git repository scanning:
      $ trivy sbom --artifact-type repo github.com/aquasecurity/trivy-ci-test

  - image archive scanning:
      $ trivy sbom --artifact-type archive ./alpine.tar
```

It won't break backwards compatibility in the sense that the only behavioural change is extra output in stdout during CI scans. I think it is acceptable as functionally theres no change. 

cc @knqyf263 - can we add a `no-progress` at a global level for Trivy so that all scan options can have it? Today it is missing in `sbom` as mentioned above.

simar7 commented on action.yaml:
```diff
@@ -20,7 +20,6 @@ inputs:
   exit-code:
     description: 'exit code when vulnerabilities were found'
     required: false
-    default: '0'
```
Good catch - I had to make this change because the default behaviour of the action is to add `--no-progress` like you mentioned but the `sbom` mode in Trivy doesn't have this flag just yet. 

```
trivy sbom --help
NAME:
   trivy sbom - generate SBOM for an artifact

USAGE:
   trivy sbom [command options] ARTIFACT

DESCRIPTION:
   ARTIFACT can be a container image, file path/directory, git repository or container image archive. See examples.

OPTIONS:
   --artifact-type value, --type value  input artifact type (image, fs, repo, archive) (default: "image") [$TRIVY_ARTIFACT_TYPE]
   --clear-cache, -c                    clear image caches without scanning (default: false) [$TRIVY_CLEAR_CACHE]
   --db-repository value                OCI repository to retrieve trivy-db from (default: "ghcr.io/aquasecurity/trivy-db") [$TRIVY_DB_REPOSITORY]
   --ignorefile value                   specify .trivyignore file (default: ".trivyignore") [$TRIVY_IGNOREFILE]
   --insecure                           allow insecure server connections when using SSL (default: false) [$TRIVY_INSECURE]
   --offline-scan                       do not issue API requests to identify dependencies (default: false) [$TRIVY_OFFLINE_SCAN]
   --output value, -o value             output file name [$TRIVY_OUTPUT]
   --sbom-format value, --format value  SBOM format (cyclonedx, spdx, spdx-json, github) (default: "cyclonedx") [$TRIVY_SBOM_FORMAT]
   --severity value, -s value           severities of vulnerabilities to be displayed (comma separated) (default: "UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL") [$TRIVY_SEVERITY]
   --skip-dirs value                    specify the directories where the traversal is skipped  (accepts multiple inputs) [$TRIVY_SKIP_DIRS]
   --skip-files value                   specify the file paths to skip traversal                (accepts multiple inputs) [$TRIVY_SKIP_FILES]
   --timeout value                      timeout (default: 5m0s) [$TRIVY_TIMEOUT]
   
EXAMPLES:
  - image scanning:
      $ trivy sbom alpine:3.15

  - filesystem scanning:
      $ trivy sbom --artifact-type fs /path/to/myapp

  - git repository scanning:
      $ trivy sbom --artifact-type repo github.com/aquasecurity/trivy-ci-test

  - image archive scanning:
      $ trivy sbom --artifact-type archive ./alpine.tar
```

It won't break backwards compatibility in the sense that the only behavioural change is extra output in stdout during CI scans. I think it is acceptable as functionally theres not change. 

cc @knqyf263 - can we add a `no-progress` at a global level for Trivy so that all scan options can have it? Today it is missing in `sbom` as mentioned above.

simar7 commented:
Sorry I don't know what is going with GitHub's inline comments at the moment. But to reply to the feedback above:

1. Added a way to submit results within the Action itself:
```
      - name: Run Trivy in GitHub SBOM mode and upload results
        uses: aquasecurity/trivy-action@master
        with:
          scan-type: 'sbom'
          format: 'github'
          output: 'dependency-results.sbom.json'
          artifact-type: 'fs'
          image-ref: '.'
          github-pat: ${{ secrets.PAT_TOKEN }}
```

2. Update README in a few places.

knqyf263 commented on README.md:
```diff
@@ -227,6 +227,39 @@ jobs:
+      - name: Run Trivy in GitHub SBOM mode and submit results to Dependency Snapshots
+        uses: aquasecurity/trivy-action@master
+        with:
+          scan-type: 'sbom'
```
We may add some breaking changes to the `sbom` subcommand. Also, it is just an alias of `image`, `fs`, etc for now. I'd suggest using `fs` now.

```suggestion
          scan-type: 'fs'
```

simar7 replied:
OK

knqyf263 commented:
>1. Added a way to submit results within the Action itself: cc @knqyf263

Looks good. Thanks!

>3. Maybe Trivy can add --no-progress as a global option as today sbom doesn't have it

The `sbom` subcommand is kind of like experimental and a bit flaky now. Can we use `fs` instead?

simar7 commented:
> > 1. Added a way to submit results within the Action itself: cc @knqyf263
> 
> Looks good. Thanks!
> 
> > 3. Maybe Trivy can add --no-progress as a global option as today sbom doesn't have it
> 
> The `sbom` subcommand is kind of like experimental and a bit flaky now. Can we use `fs` instead?

Changed it to run fs scan instead. That helped eliminate one more option of `artifact-Type` so it's more clearer now.

```
name: Pull Request
on:
  push:
    branches:
    - master
  pull_request:
jobs:
  build:
    name: Checks
    runs-on: ubuntu-20.04
    steps:
      - name: Checkout code
        uses: actions/checkout@v3

      - name: Run Trivy in GitHub SBOM mode and submit results to Dependency Snapshots
        uses: aquasecurity/trivy-action@master
        with:
          scan-type: 'fs'
          format: 'github'
          output: 'dependency-results.sbom.json'
          image-ref: '.'
          github-pat: '<github_pat_token>'
```

itaysk commented:
this is great. one small suggestion: so now if `format=github` the action will decide to automatically post the `output` to github. this is convenient but possibly too "magical" for people who don't expect it. So maybe a way to keep it convenient but less "magic" is to only post to github if specified by for example an output option like `output=github` (or `github-dependency-snapshot). just a suggestion
